### ISSUE TITLE
v14.2 - Turbo stopped working (TypeError: createContext is not a function)

### ISSUE THREAD

**Issue by nphmuller** [vercel/next.js#64400]
### Link to the code that reproduces this issue

https://github.com/nphmuller/next-14-2-turbo-api-routes-repro

### To Reproduce

- `npx create-next-app@latest`
- `npm install next-runtime-env`
- Add .env file in root with `FOO=bar`
- Add `app/api/route.ts` with this code:

```
import { env } from "next-runtime-env";

export async function GET(request: Request) {
  const foo = env("FOO");
  return new Response("Foo: " + foo);
}
```

- `npm run dev -- --turbo` (Doesn't repro on `npm run dev`
- Visit localhost:3000

### Current vs. Expected behavior

Expected: Output `Foo: bar`.
Actual:

```
TypeError: (0 , react_1.createContext) is not a function

Object.<anonymous>
node_modules/next-runtime-env/src/provider/env-context.ts (5:40)
[project]/node_modules/next-runtime-env/build/provider/env-context.js [app-route] (ecmascript)
file:///Users/nphmuller/Dev/next-14-2-turbo-api-routes-repro/.next/server/chunks/node_modules_06b366._.js (3057:3)
instantiateModule
[turbopack]/build/runtime.ts (192:19)
getOrInstantiateModuleFromParent
[turbopack]/build/runtime.ts (252:10)
commonJsRequire
[turbopack]/shared/runtime-utils.ts (240:18)
Object.<anonymous>
node_modules/next-runtime-env/src/provider/env-provider.tsx (6:1)
[project]/node_modules/next-runtime-env/build/provider/env-provider.js [app-route] (ecmascript)
file:///Users/nphmuller/Dev/next-14-2-turbo-api-routes-repro/.next/server/chunks/node_modules_06b366._.js (3085:3)
instantiateModule
[turbopack]/build/runtime.ts (192:19)
getOrInstantiateModuleFromParent
[turbopack]/build/runtime.ts (252:10)
commonJsRequire
[turbopack]/shared/runtime-utils.ts (240:18)
Object.<anonymous>
node_modules/next-runtime-env/src/provider/index.ts (3:1)
[project]/node_modules/next-runtime-env/build/provider/index.js [app-route] (ecmascript)
file:///Users/nphmuller/Dev/next-14-2-turbo-api-routes-repro/.next/server/chunks/node_modules_06b366._.js (5933:3)
instantiateModule
[turbopack]/build/runtime.ts (192:19)
getOrInstantiateModuleFromParent
[turbopack]/build/runtime.ts (252:10)
commonJsRequire
[turbopack]/shared/runtime-utils.ts (240:18)
Object.<anonymous>
node_modules/next-runtime-env/src/index.ts (10:1)
[project]/node_modules/next-runtime-env/build/index.js [app-route] (ecmascript)
file:///Users/nphmuller/Dev/next-14-2-turbo-api-routes-repro/.next/server/chunks/node_modules_06b366._.js (6292:3)
instantiateModule
[turbopack]/build/runtime.ts (192:19)
getOrInstantiateModuleFromParent
[turbopack]/build/runtime.ts (252:10)
esmImport
[turbopack]/shared/runtime-utils.ts (215:18)
```

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.4.0: Wed Feb 21 21:44:43 PST 2024; root:xnu-10063.101.15~2/RELEASE_ARM64_T6000
  Available memory (MB): 32768
  Available CPU cores: 10
Binaries:
  Node: 20.9.0
  npm: 10.2.5
  Yarn: N/A
  pnpm: 8.15.4
Relevant Packages:
  next: 14.2.1-canary.0 // Latest available version is detected (14.2.1-canary.0).
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.1.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Turbopack (--turbo)

### Which stage(s) are affected? (Select all that apply)

next dev (local)

### Additional context

This worked with Next 14.1.1.
Only happens in route handlers.
Only happens with Turbo.

<sub>[PACK-2974](https://linear.app/vercel/issue/PACK-2974/v142-turbo-stopped-working-typeerror-is-not-a-function)</sub>

**Comment by marco-elder** [comment]
I can confirm that downgrading to `~14.1.0` or turning off turbopack works, as a workaround.

**Comment by timneutkens** [comment]
Sounds quite similar to #64520, can you try next@canary?

**Comment by nphmuller** [comment]
Just tried it, and it still reproduces on 14.3.0-canary.7. I updated the repro with this version.

**Comment by nphmuller** [comment]
Looking at the issue you linked I can see what it sounds similar. Not sure if it helps, but some additional details:

I'm not directly importing a client component (that uses createContext), I'm import the `env` function from this file: https://github.com/expatfile/next-runtime-env/blob/development/src/script/env.ts

The function is re-exported via an index.ts file (https://github.com/expatfile/next-runtime-env/blob/development/src/index.ts) and that file _does_ import the code that eventually calls createContext. That code is not directly used by me.

The index.ts file contains:
> export * from './provider';

which imports/exports https://github.com/expatfile/next-runtime-env/blob/development/src/provider/env-context.ts which contains the createContext() call.

**Comment by timneutkens** [comment]
Ah I see, that's a bug in that particular library, it was a bug in Next.js that this previously didn't fail so now it's highlighting problems in libraries that are being used in Route Handlers, as Route Handlers run in the same layer as server components basically.

What needs to change in that library is that it should add `"use client"` to this file: https://github.com/expatfile/next-runtime-env/blob/development/src/provider/env-context.ts

**Comment by nphmuller** [comment]
Okay, thank you. I’ll open up a PR. Just to double check: I don’t get this error when I run ‘npm run dev’ without turbo. Does that mean this bug hasn’t been fixed in the non turbo dev build?

**Comment by nphmuller** [comment]
I just forked the project and added 'use client' to that file, and the issue still occurs. I'll double check and update the repro in a minute.

**Comment by nphmuller** [comment]
I've published the change here (on my fork): https://github.com/nphmuller/next-runtime-env/commits/fix-next-turbo-compat/

I've also published this version as @nph.muller/next-runtime-env (version 0.0.4), and updated the repro in the original post to use my fork.

**Comment by juanvilladev** [comment]
Having this same issue on 14.2.2.

Using --turbo leads to 

 ⨯ createContext only works in Client Components. Add the "use client" directive at the top of the file to use it. Read more: https://nextjs.org/docs/messages/context-in-server-component

**Comment by jmikrut** [comment]
Related:
https://github.com/vercel/next.js/issues/64412

**Comment by timneutkens** [comment]
Verified this is fixed on next@canary, similar issue to #64412 indeed.

Sandbox: https://codesandbox.io/p/github/nphmuller/next-14-2-turbo-api-routes-repro/csb-ctxpvy/draft/inspiring-mclaren?file=%2Fapp%2Fapi%2Froute.ts%3A7%2C1
Url: https://ctxpvy-3000.csb.app/api
